subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -127,12 +129,73 @@ def _get_plan_for_product_id(self, product_id: str) -> Plan:
         }
         return Plan.objects.get(**search_kwargs)
 
-    def _get_latest_transaction(self, original_transaction_id: str) -> SubscriptionPayment:
+    def _get_latest_transaction(self, original_transaction_id: str) -> Optional[SubscriptionPayment]:
+        # Latest transaction could be "CANCELLED", as it might be an upgrade.
+        # We assume that the user has a single subscription active for this app on the Apple platform.
         return SubscriptionPayment.objects.filter(
             provider_codename=self.codename,
             metadata__original_transaction_id=original_transaction_id,
         ).order_by('subscription_end').last()
 
+    def _get_active_transaction(self, transaction_id: str, original_transaction_id: str) -> SubscriptionPayment:
+        """
+        Note: this should be a simple `get`, however the database could contain duplicates.
+        Once all duplicates are defeated, replace this with `get`.
+        """
+        obtained_entries = SubscriptionPayment.objects.filter(
+            provider_codename=self.codename,
+            provider_transaction_id=transaction_id,
+            metadata__original_transaction_id=original_transaction_id,
+            status=SubscriptionPayment.Status.COMPLETED,
+        )
+        if len(obtained_entries) == 0:
+            raise SubscriptionPayment.DoesNotExist()
+
+        if len(obtained_entries) > 1:
+            logger.warning('Multiple active transactions found for transaction id "%s". '
+                           'Consider cleaning it up.', transaction_id)
```

Comment:
So overall, what did cause it to create duplicates in the database?

Reply:
The mobile app was sending 6-8 requests with the same receipt in less than 200ms. Since there was no lock on the DB and each of them was a valid receipt, in the end we had 6-8 new subscriptions.

Reply:
Replaced with `get` that should work as intended once the "remove duplicates" migration gets the job done.

Reply:
Added a try-except around it in case the measures taken in this PR still fail.